File: backend/app/api/v1/endpoints/employees.py
import uuid

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.core.deps import get_current_company_id, get_current_user, get_db, require_permission
from app.models.user import User
from app.schemas.common import Page
from app.schemas.employee import (
    EmployeeCreateRequest,
    EmployeeDetailResponse,
    EmployeeSelfUpdateRequest,
    EmployeeSummaryResponse,
    EmployeeUpdateRequest,
)
from app.services.employee_service import employee_service

router = APIRouter(prefix="/employees", tags=["employees"])


@router.get("", response_model=Page[EmployeeSummaryResponse])
def list_employees(
    search: str | None = Query(default=None),
    department_id: uuid.UUID | None = Query(default=None),
    status: str | None = Query(default=None),
    manager_id: uuid.UUID | None = Query(default=None),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=25, ge=1, le=100),
    company_id: uuid.UUID = Depends(get_current_company_id),
    db: Session = Depends(get_db),
    _: User = Depends(require_permission("employee", "view")),
):
    return employee_service.search_employees(
        db,
        company_id,
        search=search,
        department_id=department_id,
        status=status,
        manager_id=manager_id,
        page=page,
        page_size=page_size,
    )


@router.post("", response_model=EmployeeDetailResponse, status_code=201)
def create_employee(
    payload: EmployeeCreateRequest,
    company_id: uuid.UUID = Depends(get_current_company_id),
    db: Session = Depends(get_db),
    current_user: User = Depends(require_permission("employee", "create")),
):
    return employee_service.create_employee(
        db,
        company_id,
        email=payload.email,
        first_name=payload.first_name,
        last_name=payload.last_name,
        phone=payload.phone,
        department_id=payload.department_id,
        designation=payload.designation,
        manager_id=payload.manager_id,
        employment_type=payload.employment_type,
        joining_date=payload.joining_date,
        role_ids=payload.role_ids,
        actor_user_id=current_user.id,
    )


@router.get("/me", response_model=EmployeeDetailResponse)
def get_my_profile(current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return employee_service.get_employee_by_user_id(db, current_user.id)


@router.patch("/me", response_model=EmployeeDetailResponse)
def update_my_profile(
    payload: EmployeeSelfUpdateRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    employee = employee_service.get_employee_by_user_id(db, current_user.id)
    return employee_service.update_self(db, current_user.company_id, employee.id, phone=payload.phone)


@router.get("/{employee_id}", response_model=EmployeeDetailResponse)
def get_employee(
    employee_id: uuid.UUID,
    company_id: uuid.UUID = Depends(get_current_company_id),
    db: Session = Depends(get_db),
    _: User = Depends(require_permission("employee", "view")),
):
    return employee_service.get_employee(db, company_id, employee_id)


@router.patch("/{employee_id}", response_model=EmployeeDetailResponse)
def update_employee(
    employee_id: uuid.UUID,
    payload: EmployeeUpdateRequest,
    company_id: uuid.UUID = Depends(get_current_company_id),
    db: Session = Depends(get_db),
    current_user: User = Depends(require_permission("employee", "update")),
):
    updates = payload.model_dump(exclude_unset=True)
    return employee_service.update_employee(db, company_id, employee_id, actor_user_id=current_user.id, **updates)


@router.delete("/{employee_id}", status_code=204)
def delete_employee(
    employee_id: uuid.UUID,
    company_id: uuid.UUID = Depends(get_current_company_id),
    db: Session = Depends(get_db),
    current_user: User = Depends(require_permission("employee", "delete")),
):
    employee_service.delete_employee(db, company_id, employee_id, actor_user_id=current_user.id)

File: frontend/src/api/employees.ts
import { apiClient } from "@/api/client";
import type { EmployeeDetail, EmployeeSummary, Page } from "@/types";
import { triggerCsvDownload } from "@/utils/downloadCsv";

export interface EmployeeListFilters {
  search?: string;
  department_id?: string;
  status?: string;
  page: number;
  page_size: number;
}

export interface EmployeeInput {
  email: string;
  first_name: string;
  last_name: string;
  phone?: string | null;
  department_id?: string | null;
  designation?: string | null;
  manager_id?: string | null;
  employment_type: string;
  location?: string | null;
  joining_date?: string | null;
  role_ids?: string[];
}

export type EmployeeUpdateInput = Partial<Omit<EmployeeInput, "email" | "role_ids">> & { status?: string };

export async function listEmployees(filters: EmployeeListFilters): Promise<Page<EmployeeSummary>> {
  const response = await apiClient.get<Page<EmployeeSummary>>("/employees", { params: filters });
  return response.data;
}

export async function getEmployee(id: string): Promise<EmployeeDetail> {
  const response = await apiClient.get<EmployeeDetail>(`/employees/${id}`);
  return response.data;
}

export async function createEmployee(payload: EmployeeInput): Promise<EmployeeDetail> {
  const response = await apiClient.post<EmployeeDetail>("/employees", payload);
  return response.data;
}

export async function updateEmployee(id: string, payload: EmployeeUpdateInput): Promise<EmployeeDetail> {
  const response = await apiClient.patch<EmployeeDetail>(`/employees/${id}`, payload);
  return response.data;
}

export async function deleteEmployee(id: string): Promise<void> {
  await apiClient.delete(`/employees/${id}`);
}

export async function setEmployeeRoles(id: string, roleIds: string[]): Promise<EmployeeDetail> {
  const response = await apiClient.put<EmployeeDetail>(`/employees/${id}/roles`, { role_ids: roleIds });
  return response.data;
}

export async function getMyProfile(): Promise<EmployeeDetail> {
  const response = await apiClient.get<EmployeeDetail>("/employees/me");
  return response.data;
}

export interface EmployeeExportFilters {
  search?: string;
  departmentId?: string;
  status?: string;
  managerId?: string;
  employmentType?: string;
  location?: string;
  joiningDateFrom?: string;
  joiningDateTo?: string;
}

export async function downloadEmployeesExportCsv(filters: EmployeeExportFilters = {}): Promise<void> {
  const response = await apiClient.get<Blob>("/employees/export", {
    params: {
      search: filters.search,
      department_id: filters.departmentId,
      status: filters.status,
      manager_id: filters.managerId,
      employment_type: filters.employmentType,
      location: filters.location,
      joining_date_from: filters.joiningDateFrom,
      joining_date_to: filters.joiningDateTo,
    },
    responseType: "blob",
  });
  triggerCsvDownload(response.data, "employees-export.csv");
}
